### DISCUSSION TITLE
Ecma script version in iTowns

### DISCUSSION THREAD

**Discussion by gchoqueux** [iTowns/itowns#1920] [General]
## Ecma script 6 in **itowns** 
**ITowns** is written in **ES6** and uses almost all its new features. ([See all features](http://es6-features.org)).
**ITowns** is transpiled in **ES5** with [Babel](https://babeljs.io/).

### Important code rules, that aren't completely applied:
- [ ] Uses `Class` instead of object.
- [ ] Uses `Set` and `Map` instead of `Array`, when relevant.
### Little used features, don't forget them :
- [ ] Array element finding
```js
[ 1, 3, 4, 2 ].find(x => x > 3) // 4 
[ 1, 3, 4, 2 ].findIndex(x => x > 3) // 2
```
- [ ] String Searching
```js
"hello".startsWith("ello", 1) // true 
"hello".endsWith("hell", 4)  // true 
"hello".includes("ell")  // true
```
- [ ] Number Safety Checking
```js
Number.isSafeInteger(9007199254740992) === false
```
## Ecma script 2016 in **itowns** see #932 
- [x] `exponentiation operator (**)`.
- [x] `Array.prototype.includes`.

## Ecma script 2017
- [ ] Includes `async/await`. (Used in tests)

## Ecma script 2018
- [ ] New features include `rest/spread properties`.
- [ ] Asynchronous iteration.
- [ ] `Promise.prototype.finally()` .
- [ ] additions to RegExp.

## Ecma script 2019
[What's new in Ecma script 2019](https://medium.com/@selvaganesh93/javascript-whats-new-in-ecmascript-2019-es2019-es10-35210c6e7f4b)

- [ ] `Array.Flat()`
- [ ] `Object.fromEntries()`
- [ ] `Array.flatMap()`
- [ ] `String.trimStart() & String.trimEnd()`
- [ ] Optional Catch Binding
- [ ] Function.toString()
- [ ] Well Formed JSON.Stringify()
- [ ] Array.Sort Stability

## Ecma script 2020
[What's new in Ecma script 2020](https://auth0.com/blog/javascript-whats-new-es2020/)

- [ ] Dynamic imports
- [ ] Import meta data
- [ ] New export syntax
- [ ] `BigInt` and arbitrary precision integers
- [ ] `matchAll() `method for regular expressions
- [ ] `globalThis` object
- [ ] `Promise.allSettled()` method
- [ ] nullish coalescing operator
- [ ] Optional chaining

## Ecma script 2021
[What's new in Ecma script 2021](https://dev.to/faithfulojebiyi/new-features-in-ecmascript-2021-with-code-examples-302h)

- [ ] `string.replaceAll` method
- [ ] Private Methods
- [ ] Private Accessors 
- [ ] `Promise.any()`
- [ ] Numeric Separators 
- [ ] Logical Assignment Operator

## Ecma script 2022
[What's new in Ecma script 2022](https://widgetcore.com/new-features-to-be-included-in-javascript-after-es-2022/)

- [ ] Top-level Await Operator
- [ ] Class Field Declarations
- [ ] Private Methods and Fields
- [ ] Static Class Fields and Private Static Methods
- [ ] Regexp Match Indices
- [ ] Ergonomic Brand Checks for Private Fields
- [ ]  `Array.at()` Function for Indexing
- [ ] Scalable `Object.prototype.hasOwnProperty()`

**Comment by jailln** [comment]
Good idea.

I think that some of this information (e.g. use class instead of object, etc.) should be added in a to be created 'developer guide' (maybe in the wiki or in the docs folder? and pointed into the CONTRIBUTING.md file).

**Comment by zarov** [comment]
Thanks for creating this issue ! This can be a good first contribution for anyone wanting to contribute to iTowns.

@jailln yes, in fact the `CONTRIBUTING` file could be rewritten for a more up-to-date content. If you want to start a draft I'm all for it and then we can discuss it in a PR ?

**Comment by jailln** [comment]
@zarov yep, I'll start a draft in a new PR with propositions that came up while starting contributing to iTowns.

**Comment by mgermerie** [comment]
We need to define what rule we should make mandatory (in the ESLint config).
